### PR TITLE
support new marian models

### LINKED ISSUE #15109: Why is Marian to Torch converter hardcoded for tied vocab ?
I see the following condition:
https://github.com/huggingface/transformers/blob/16f0b7d72c6d4e122957392c342b074aa2c5c519/src/transformers/models/marian/convert_marian_to_pytorch.py#L462

While training my Marian model, I do not want to tie my source and target embeddings.
How do I convert such a model? 

I see that in `MarianConfig` itself, this is not supported:
https://github.com/huggingface/transformers/blob/16f0b7d72c6d4e122957392c342b074aa2c5c519/src/transformers/models/marian/configuration_marian.py#L46-L49

Can this be considered a **feature request** to make it generic?

---

Also, why is the `hidden-dim` required to be `512` in the converter?
https://github.com/huggingface/transformers/blob/16f0b7d72c6d4e122957392c342b074aa2c5c519/src/transformers/models/marian/convert_marian_to_pytorch.py#L478

What if I train transformer-big models?

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR updates the Marian model:
1. To allow not sharing embeddings between encoder and decoder.
2. Allow tying only decoder embeddings with `lm_head`.
3. Separate two `vocabs` in `tokenizer` for `src` and `tgt` language

To support this, the PR introduces the following new methods:
- `get_decoder_input_embeddings` and `set_decoder_input_embeddings`
To get and set the decoder embeddings when the embeddings are not shared. These methods will raise an error if the embeddings are shared.
- `resize_decoder_token_embeddings`
To only resize the decoder embeddings. Will raise an error if the embeddings are shared.

This PR also adds two new config attributes to `MarianConfig`:
- `share_encoder_decoder_embeddings`: to indicate if emb should be shared or not
- `decoder_vocab_size`: to specify the vocab size for decoder when emb are not shared.

And the following methods from `PreTrainedModel` class are overridden to support these changes:
- `tie_weights`
- `_resize_token_embeddings`

Fixes #15109

### CODE REVIEW

File: src/transformers/models/marian/modeling_marian.py [hunk 1]
```diff
@@ -1239,9 +1272,60 @@ def get_decoder(self):
...
+
+        # if word embeddings are not tied, make sure that lm head is resized as well
+        if (
+            self.config.share_encoder_decoder_embeddings
+            and self.get_output_embeddings() is not None
+            and not self.config.tie_word_embeddings
+        ):
+            old_lm_head = self.get_output_embeddings()
+            new_lm_head = self._get_resized_lm_head(old_lm_head, new_num_tokens)
+            self.set_output_embeddings(new_lm_head)
```
Comment (patil-suraj) [inline src/transformers/models/marian/modeling_marian.py:1293]: This will only resize the `lm_head` if embeddings are shared.

File: src/transformers/models/marian/modeling_marian.py [hunk 2]
```diff
@@ -1257,6 +1341,28 @@ def get_output_embeddings(self):
...
+        Tie the weights between the input embeddings and the output embeddings.
+
+        If the `torchscript` flag is set in the configuration, can't handle parameter sharing so we are cloning the
+        weights instead.
+        """
+        output_embeddings = self.get_output_embeddings()
+        if output_embeddings is not None and getattr(self.config, "tie_word_embeddings", True):
+            # if embeddings are shared this will return shared embeddings otherwise decoder embed_tokens
+            word_embeddings = self.get_decoder().get_input_embeddings()
+            self._tie_or_clone_weights(output_embeddings, word_embeddings)
```
Comment (patil-suraj) [inline src/transformers/models/marian/modeling_marian.py:1355]: We always return decoder embeddings here. This should work for both cases, shared or not shared.

File: src/transformers/models/marian/modeling_marian.py [hunk 3]
```diff
@@ -823,7 +829,8 @@ def __init__(self, config: MarianConfig, embed_tokens: Optional[nn.Embedding] =
         if embed_tokens is not None:
             self.embed_tokens = embed_tokens
         else:
-            self.embed_tokens = nn.Embedding(config.vocab_size, config.d_model, self.padding_idx)
+            # TODO: can we assume if embeddings are not shared the vocabs are also not shared ?
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:832]: I think so yes. Also since we let `config.decoder_vocab_size` default to `config.vocab_size` I don't think that this is a problem no?

File: src/transformers/models/marian/modeling_marian.py [hunk 4]
```diff
@@ -1092,12 +1103,32 @@ def __init__(self, config: MarianConfig):
...
+        if self.config.share_encoder_decoder_embeddings:
+            self.shared = value
+            self.encoder.embed_tokens = self.shared
+            self.decoder.embed_tokens = self.shared
+        else:  # if not shared only set encoder embeedings
+            self.encoder.embed_tokens = value
+
+    def get_decoder_input_embeddings(self):
+        if self.config.share_encoder_decoder_embeddings:
+            raise ValueError(
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1119]: Why raise an error here? It's totally fine to just return `self.get_input_embeddigs()` in this case no?

File: src/transformers/models/marian/modeling_marian.py [hunk 5]
```diff
@@ -1092,12 +1103,32 @@ def __init__(self, config: MarianConfig):
...
+                "`get_decoder_input_embeddings` should not be called if `config.share_encoder_decoder_embeddings` "
+                "is `True`. Please use `get_input_embeddings` instead."
+            )
+        return self.get_decoder().get_input_embeddings()
+
+    def set_decoder_input_embeddings(self, value):
+        if self.config.share_encoder_decoder_embeddings:
+            raise ValueError(
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1127]: Here an error makes much more sense to me. But I think we can be a bit more explicit. E.g. "config.share_encoder_decoder_embeddings is set to True meaning the decoder input embeddings are shared with the encoder. In order to set the decoder input embeddings, you should simply set the encoder input embeddings by calling ..."

File: src/transformers/models/marian/modeling_marian.py [hunk 6]
```diff
@@ -1257,6 +1341,28 @@ def get_output_embeddings(self):
     def set_output_embeddings(self, new_embeddings):
         self.lm_head = new_embeddings
 
+    def tie_weights(self):
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1344]: ok for me

File: src/transformers/models/marian/modeling_marian.py [hunk 7]
```diff
@@ -1225,8 +1256,10 @@ class MarianMTModel(MarianPreTrainedModel):
     def __init__(self, config: MarianConfig):
         super().__init__(config)
         self.model = MarianModel(config)
-        self.register_buffer("final_logits_bias", torch.zeros((1, self.model.shared.num_embeddings)))
-        self.lm_head = nn.Linear(config.d_model, self.model.shared.num_embeddings, bias=False)
+
+        self.output_size = config.vocab_size if config.share_encoder_decoder_embeddings else config.decoder_vocab_size
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1260]: I'd rather call it `self.output_vocab_size` or `self.target_vocab_size`

**Review by patrickvonplaten — APPROVED** [review]
Overall, I'm in favor of adding the new Marian checkpoints the way it is shown here. The change from having a Marian model that always **force-tied** encoder and decoder embeddings to a Marian model that can switch between **force-tied** and no-tied encoder input embeddings and encoder output embeddings is the better option here IMO even though it does go a bit again our philosophy of not changing existing model code. 

The main reasons why I'm in favor of the approach as it's implemented now are (with the feedback given below):
- All the changes of this PR are also applicable to existing Marian V1 checkpoints. More specifically all Marian V1 checkpoints can be loaded here with `share_encoder_decoder_embeddings=False` and then fine-tuned with embeddings not being tied.
- Marian V2 comes from the exact same library as Marian V1 and is the same model. Creating a new name here (Marian V2) could confuse users.

Thoughts @LysandreJik @sgugger ?

File: src/transformers/models/marian/modeling_marian.py [hunk 8]
```diff
@@ -1083,7 +1090,11 @@ def __init__(self, config: MarianConfig):
         super().__init__(config)
 
         padding_idx, vocab_size = config.pad_token_id, config.vocab_size
-        self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
+
+        if config.share_encoder_decoder_embeddings:
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1094]: As discussed offline, I think we should rewrite this a bit. Even if `config.share_encoder_decoder_embeddings is False` we should clone those weights to initialized encoder and decoder with it. This makes this PR much more appealing for existing Marian models. E.g. one could then simple load a Marian model with:


```python
from transformers import MarianMTModel

model = MarianMTModel.from_pretrained("Helsinki-NLP/opus-mt-en-de", share_encoder_decoder_embeddings=False)

# now one can fine-tune the model with encoder / decoder embeddings not being shared which has been shown to lead to improved results
```

File: src/transformers/models/marian/convert_marian_to_pytorch.py [hunk 9]
```diff
@@ -398,7 +418,7 @@ def check_equal(marian_cfg, k1, k2):
 
 def check_marian_cfg_assumptions(marian_cfg):
     assumed_settings = {
-        "tied-embeddings-all": True,
+        # "tied-embeddings-all": True,
```
Comment (sgugger) [inline src/transformers/models/marian/convert_marian_to_pytorch.py:421]: To be cleaned up?

File: src/transformers/models/marian/convert_marian_to_pytorch.py [hunk 10]
```diff
@@ -417,9 +437,9 @@ def check_marian_cfg_assumptions(marian_cfg):
         actual = marian_cfg[k]
         if actual != v:
             raise ValueError(f"Unexpected config value for {k} expected {v} got {actual}")
-    check_equal(marian_cfg, "transformer-ffn-activation", "transformer-aan-activation")
-    check_equal(marian_cfg, "transformer-ffn-depth", "transformer-aan-depth")
-    check_equal(marian_cfg, "transformer-dim-ffn", "transformer-dim-aan")
+    # check_equal(marian_cfg, "transformer-ffn-activation", "transformer-aan-activation")
+    # check_equal(marian_cfg, "transformer-ffn-depth", "transformer-aan-depth")
+    # check_equal(marian_cfg, "transformer-dim-ffn", "transformer-dim-aan")
```
Comment (sgugger) [inline src/transformers/models/marian/convert_marian_to_pytorch.py:442]: To be cleaned up?

File: src/transformers/models/marian/convert_marian_to_pytorch.py [hunk 11]
```diff
@@ -475,14 +507,16 @@ def __init__(self, source_dir, eos_token_id=0):
         self.source_dir = source_dir
         self.cfg = cfg
         hidden_size, intermediate_shape = self.state_dict["encoder_l1_ffn_W1"].shape
-        if hidden_size != 512 or cfg["dim-emb"] != 512:
-            raise ValueError(f"Hidden size {hidden_size} and configured size {cfg['dim_emb']} mismatched or not 512")
+        # if hidden_size != 512 or cfg["dim-emb"] != 512:
+        #     raise ValueError(f"Hidden size {hidden_size} and configured size {cfg['dim_emb']} mismatched or not 512")
```
Comment (sgugger) [inline src/transformers/models/marian/convert_marian_to_pytorch.py:511]: Same here

File: src/transformers/models/marian/convert_marian_to_pytorch.py [hunk 12]
```diff
@@ -592,16 +639,24 @@ def convert(source_dir: Path, dest_dir):
     dest_dir = Path(dest_dir)
     dest_dir.mkdir(exist_ok=True)
 
-    add_special_tokens_to_vocab(source_dir)
-    tokenizer = MarianTokenizer.from_pretrained(str(source_dir))
-    tokenizer.save_pretrained(dest_dir)
+    # add_special_tokens_to_vocab(source_dir)
+    # tokenizer = MarianTokenizer.from_pretrained(str(source_dir))
+    # tokenizer.save_pretrained(dest_dir)
```
Comment (sgugger) [inline src/transformers/models/marian/convert_marian_to_pytorch.py:644]: To be cleaned up?

File: src/transformers/models/marian/convert_marian_to_pytorch.py [hunk 13]
```diff
@@ -592,16 +639,24 @@ def convert(source_dir: Path, dest_dir):
...
 
     # retrieve EOS token and set correctly
-    tokenizer_has_eos_token_id = hasattr(tokenizer, "eos_token_id") and tokenizer.eos_token_id is not None
-    eos_token_id = tokenizer.eos_token_id if tokenizer_has_eos_token_id else 0
+    # tokenizer_has_eos_token_id = hasattr(tokenizer, "eos_token_id") and tokenizer.eos_token_id is not None
+    # eos_token_id = tokenizer.eos_token_id if tokenizer_has_eos_token_id else 0
+    # TODO: ^^
```
Comment (sgugger) [inline src/transformers/models/marian/convert_marian_to_pytorch.py:649]: What's the TODO, don't forget ;-)

**Review by sgugger — APPROVED** [review]
Ok for me. It's really pushing the test for a new model to its limit, but I understand the arguments to keep it in the same model.

File: src/transformers/models/marian/modeling_marian.py [hunk 14]
```diff
@@ -1085,26 +1092,78 @@ def __init__(self, config: MarianConfig):
         padding_idx, vocab_size = config.pad_token_id, config.vocab_size
         self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
 
-        self.encoder = MarianEncoder(config, self.shared)
-        self.decoder = MarianDecoder(config, self.shared)
+        if self.config.share_encoder_decoder_embeddings:
+            encoder_embed_tokens = decoder_embed_tokens = self.shared
+        else:
+            encoder_embed_tokens = copy.deepcopy(self.shared)
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1098]: Can we maybe add a nice comment here why we do the cloning in this branch and why we don't do it above? E.g. if it's' not shared we want the embed_tokens to be correctly initialized, but **not** tied which is why we do the deepcopy here. Think it's worth adding a comment here

File: src/transformers/models/marian/modeling_marian.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/models/marian/modeling_marian.py:1122]: Still don't think we need to raise here ;-)

File: src/transformers/models/marian/modeling_marian.py [hunk 15]
```diff
@@ -1239,9 +1302,60 @@ def get_decoder(self):
 
     def resize_token_embeddings(self, new_num_tokens: int) -> nn.Embedding:
         new_embeddings = super().resize_token_embeddings(new_num_tokens)
-        self._resize_final_logits_bias(new_num_tokens)
+        if self.config.share_encoder_decoder_embeddings:
+            self._resize_final_logits_bias(new_num_tokens)
         return new_embeddings
 
+    def _resize_token_embeddings(self, new_num_tokens):
+        # TODO: override this method to handle shared not shared
```
Comment (patrickvonplaten) [inline src/transformers/models/marian/modeling_marian.py:1310]: remove TODO?

**Review by patrickvonplaten — COMMENTED** [review]
Looks good to me in general.
Left a couple of comments. 

Also given that a bunch of new model checkpoints will be added here - let's maybe add a slow integration test as well?
